Index: content-review-api/public/src/java/org/sakaiproject/contentreview/service/ContentReviewService.java
===================================================================
--- content-review-api/public/src/java/org/sakaiproject/contentreview/service/ContentReviewService.java	(revision 308340)
+++ content-review-api/public/src/java/org/sakaiproject/contentreview/service/ContentReviewService.java	(working copy)
@@ -51,8 +51,7 @@
 	 *  @param Reference to the content object that should be submitted
 	 *  
 	 */
-	public void queueContent(String userId, String siteId, String taskId, String contentId)
-	throws QueueException;
+	public void queueContent(String userId, String siteId, String assignmentReference, List<String> contentIds, String inlineText, long submissionTime) throws QueueException;
 	
 	/**
 	 *  Retrieve a score for a specific item
@@ -64,37 +63,47 @@
 	  */
 	public int getReviewScore(String contentId)
 	throws QueueException, ReportException, Exception;
+
+	public int getReviewScore(String contentId, String assignmentRef, String userId) throws QueueException,
+                        ReportException, Exception;
+
+	public String getInlineTextId(String assignmentReference, String userId, long submissionTime);
 	
+	public boolean acceptInlineAndMultipleAttachments();
+
 	/**
 	 *  Get the URL of the report
 	 * @param contentId
+	 * @param assignmentRef
 	 * @return the url
 	 * @throws QueueException
 	 * @throws ReportException
 	 * * * @deprecated since Nov 2007, use {@link getReviewReportInstructor(String contentId)} or {@link getReviewReportInstructor(String contentId)}
 	 */
-	public String getReviewReport(String contentId)
+	public String getReviewReport(String contentId, String assignmentRef)
 	throws QueueException, ReportException;
 	
 	/**
 	 * Get the URL of a report constructed for a student
 	 * @param contentId
+	 # @param assignmentRef
 	 * @return
 	 * @throws QueueException
 	 * @throws ReportException
 	 * */
-	public String getReviewReportStudent(String contentId)
+	public String getReviewReportStudent(String contentId, String assignmentRef)
 	throws QueueException, ReportException;
 	
 	/**
 	 * Get the URL for a report constructed for an Instructor
 	 * 
 	 * @param contentId
+	 * @param assignmentRef
 	 * @return
 	 * @throws QueueException
 	 * @throws ReportException
 	 */
-	public String getReviewReportInstructor(String contentId)
+	public String getReviewReportInstructor(String contentId, String assignmentRef)
 	throws QueueException, ReportException;
 	
 	
